### PR TITLE
opentofu: init at v1.6.0-alpha1

### LINKED ISSUE #253469: Package request: opentofu
**Project description**

OpenTofu is a fork of Terraform, which is no longer under a free license (and should be marked as unfree for releases >= 1.6.0). This fork is from the free sources before the license change to a source-available BSL license.

**Metadata**

* homepage URL: https://opentofu.org/
* source URL: https://github.com/opentofu/opentofu
* license: mpl
* platforms: unix

### PR DESCRIPTION
**Pull request by gmemstr**
## Description of changes

OpenTofu is a fork of Terraform pre-BSL, so the package is pretty similar to the Terraform one with names swapped around. Happy to take this further if needed but given the current state of opentofu I felt this was probably? okay. That said, I did remove the existing Terraform maintainers as well, but am happy to re-add them if they're good with maintaining opentofu as well.

~~Opentofu is not yet tagged with a version so I've opted to go with the latest commit at the time of creation.~~

Closes https://github.com/NixOS/nixpkgs/issues/253469

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by dylanmtaylor — APPROVED** [review]

**Comment by gmemstr** [comment]
> It seems that opentofu can be packaged with terraform.overrideAttrs ...? That way, maintainers don't have to review hundreds of lines of code.

The expectation I have is that opentofu is going to diverge from Terraform in a way that means the overrides added would end up making more sense as  a new package/splitting off. I'm not against merging the two but I'm a little cautious about this.

> Probably tag some of them in this PR?

I left a comment in the associated issue that should hopefully reach their inboxes, but I may as well pull them in here as well @marsam @zowoq @techknowlogick

> Use version = "unstable-2023-XX-XX", according to [pkgs/README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md?rgh-link-date=2023-09-20T15%3A30%3A56Z#package-naming).

Done! :)

**Review by sugar700 — CHANGES_REQUESTED** [review]
I'm not a fan of the current derivation name `opentofu.opentofu_unstable`. I don't think it's intended, and the name is this long because `callPackage` accesses an object that contains `opentofu_unstable` field. Could we simplify that to just `opentofu`?

**Comment by xfix** [comment]
Terraform package mostly did that because in the past we had multiple versions of Terraform, but now that we only have Terraform 1, this is no longer necessary due to [Terraform v1 compatibility promises](https://developer.hashicorp.com/terraform/language/v1-compatibility-promises).

That said, I have no idea what sort of compatibility promises OpenTofu will have, but I think that major breaking changes in a popular programming language are pretty universally understood as a bad idea after Python 3.

I think `opentofu` as a package name (with no namespacing, not `opentofu.opentofu`) sounds right to me, and if branching will be necessary, we could call packages `opentofu_2` or something like that.

Note that `terraform`/`terraform_1` is defined as `(callPackage ../applications/networking/cluster/terraform { }).terraform_1`.

**Comment by gmemstr** [comment]
Took me a little bit but managed to get the package out of the namespace and ready for another review.

**Review by dylanmtaylor — APPROVED** [review]

File: pkgs/applications/networking/cluster/opentofu/default.nix [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+        export TF_SKIP_REMOTE_TESTS=1
+      '';
+
+    subPackages = [ "." ];
+
+    meta = with lib; {
+      description =
+        "Tool for building, changing, and versioning infrastructure";
+      homepage = "https://opentofu.org/";
+      changelog = "https://github.com/opentofu/opentofu/blob/${version}/CHANGELOG.md";
```
Comment (cafkafk) [inline pkgs/applications/networking/cluster/opentofu/default.nix:58]: Currently, I'm pretty sure this evaluates to https://github.com/opentofu/opentofu/blob/unstable-2023-09-20/CHANGELOG.md

It probably should be the `src.revision` instead https://github.com/opentofu/opentofu/blob/922f61d84742804dc4357a90ae56c20a07a80683/CHANGELOG.md

**Review by cafkafk — CHANGES_REQUESTED** [review]
Hey, really cool that you're doing this, can't wait to have this available.

**Comment by gmemstr** [comment]
Bumped the revision and addressed comments from @cafkafk. Also noticed that the way to build the binary changed slightly and the main binary's name changed so updated that.

**Review by adamcstephens — APPROVED** [review]
Result of `nixpkgs-review pr 256307` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>opentofu</li>
  </ul>
</details>

-----

Result of `nixpkgs-review pr 256307` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>opentofu</li>
  </ul>
</details>

-----

Result of `nixpkgs-review pr 256307` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>opentofu</li>
  </ul>
</details>

-----

Result of `nixpkgs-review pr 256307` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>opentofu</li>
  </ul>
</details>

**Comment by azahi** [comment]
Result of `nixpkgs-review pr 256307` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>opentofu</li>
  </ul>
</details>

**Comment by azahi** [comment]
@ofbord build opentofu opentofu.passthru.tests

**Comment by gmemstr** [comment]
Noted the failures and updated the binary name in the wrapping process.

**Comment by gmemstr** [comment]
@ofborg build opentofu opentofu.passthru.tests

**Comment by jocelynthode** [comment]
Thanks  for your work @gmemstr! OpenTofu finally released their first version: https://github.com/opentofu/opentofu/releases/tag/v1.6.0-alpha1  might be worth it to switch to tags from now on :)

**Comment by gmemstr** [comment]
Dangit, of course they did <:) PR updated

**Review by adamcstephens — APPROVED** [review]
Let's get this one merged so future bumps should be quicker. :)

---

Result of `nixpkgs-review pr 256307` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>opentofu</li>
  </ul>
</details>

-----

Result of `nixpkgs-review pr 256307` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>opentofu</li>
  </ul>
</details>

-----

Result of `nixpkgs-review pr 256307` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>opentofu</li>
  </ul>
</details>

-----

Result of `nixpkgs-review pr 256307` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>opentofu</li>
  </ul>
</details>

**Review by dylanmtaylor — APPROVED** [review]
I agree with Adam. We should definitely get an initial version in.

**Comment by azahi** [comment]
Looks like the tests are now failing because `registry.opentofu.org` is not accessible.

**Comment by gmemstr** [comment]
@azahi yeah not sure what's up with that or how we want to proceed, whether it be removing the tests or ignoring them for the time being.

**Comment by gmemstr** [comment]
Unless I messed something up while moving around the tests and their inputs. Could be related to that but not 100% sure where I went wrong.

**Comment by azahi** [comment]
> @azahi yeah not sure what's up with that or how we want to proceed, whether it be removing the tests or ignoring them for the time being.

The PR looks good and I also want to see this merged. I guess we can just disable checks for now and leave a comment explaining the reasoning. The most important part is to not to forget about all of that later :)

> Could be related to that

I think they broke something on their end. It just returns "Missing Authentication Token".

File: pkgs/applications/networking/cluster/opentofu/default.nix [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+let
+  package =  buildGoModule rec {
+    pname = "opentofu";
+    version = "v1.6.0-alpha1";
+
+    src = fetchFromGitHub {
+      owner = "opentofu";
+      repo = "opentofu";
+      rev = "${version}";
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/cluster/opentofu/default.nix:22]:
```suggestion
    version = "1.6.0-alpha1";

    src = fetchFromGitHub {
      owner = "opentofu";
      repo = "opentofu";
      rev = "v${version}";
```

File: pkgs/applications/networking/cluster/opentofu/default.nix [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+      hash = "sha256-0FO55H1nOyhAd+ex1zA0XycH6x/HKkLlxzuIJNoaI9g=";
+    };
+    vendorHash = "sha256-3jQfIIZOgOmNHQ06rXz+3QTZ37WcuCc7A7/MhC7udrg=";
+    ldflags = [ "-s" "-w" ];
+
+    postConfigure = ''
+        # speakeasy hardcodes /bin/stty https://github.com/bgentry/speakeasy/issues/22
+        substituteInPlace vendor/github.com/bgentry/speakeasy/speakeasy_unix.go \
+          --replace "/bin/stty" "${coreutils}/bin/stty"
+      '';
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/cluster/opentofu/default.nix:32]:
```suggestion
      # speakeasy hardcodes /bin/stty https://github.com/bgentry/speakeasy/issues/22
      substituteInPlace vendor/github.com/bgentry/speakeasy/speakeasy_unix.go \
        --replace "/bin/stty" "${coreutils}/bin/stty"
    '';
```
I am surprised that go is not complaining about this

File: pkgs/applications/networking/cluster/opentofu/default.nix [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+    # https://github.com/posener/complete/blob/9a4745ac49b29530e07dc2581745a218b646b7a3/cmd/install/bash.go#L8
+    postInstall = ''
+        installShellCompletion --bash --name tofu <(echo complete -C tofu tofu)
+      '';
+
+    preCheck = ''
+        export HOME=$TMPDIR
+        export TF_SKIP_REMOTE_TESTS=1
+      '';
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/cluster/opentofu/default.nix:50]:
```suggestion
      installShellCompletion --bash --name tofu <(echo complete -C tofu tofu)
    '';

    preCheck = ''
      export HOME=$TMPDIR
      export TF_SKIP_REMOTE_TESTS=1
    '';
```

File: pkgs/applications/networking/cluster/opentofu/default.nix [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+    subPackages = [ "./cmd/..." ];
+
+    meta = with lib; {
+      description =
+        "Tool for building, changing, and versioning infrastructure";
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/cluster/opentofu/default.nix:56]:
```suggestion
      description = "Tool for building, changing, and versioning infrastructure";
```

File: pkgs/applications/networking/cluster/opentofu/default.nix [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+        gmemstr
+      ];
+      mainProgram = "tofu";
+    };
+  };
+
+  opentofu_plugins_test = let
+    mainTf = writeText "main.tf" ''
+        resource "random_id" "test" {}
+      '';
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/cluster/opentofu/default.nix:70]:
```suggestion
      resource "random_id" "test" {}
    '';
```

File: pkgs/applications/networking/cluster/opentofu/default.nix [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+    opentofu = package.withPlugins (p: [ p.random ]);
+    test =
+      runCommand "opentofu-plugin-test" { buildInputs = [ opentofu ]; } ''
+          set -e
+          # make it fail outside of sandbox
+          export HTTP_PROXY=http://127.0.0.1:0 HTTPS_PROXY=https://127.0.0.1:0
+          cp ${mainTf} main.tf
+          tofu init
+          touch $out
+        '';
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/cluster/opentofu/default.nix:80]:
```suggestion
    test = runCommand "opentofu-plugin-test" {
      buildInputs = [ opentofu ];
    } ''
      # make it fail outside of sandbox
      export HTTP_PROXY=http://127.0.0.1:0 HTTPS_PROXY=https://127.0.0.1:0
      cp ${mainTf} main.tf
      tofu init
      touch $out
    '';
```

set -e should already be set here

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
I only correct the first few indentations which where 4 instead of 2 spaces.

**Review by marsam — APPROVED** [review]

**Pull request merged**